importer/balanced/builder.go

```diff
@@ -1,24 +1,40 @@
-// Package balanced provides methods to build balanced DAGs.
-// In a balanced DAG, nodes are added to a single root
-// until the maximum number of links is reached (with leaves
-// being at depth 0). Then, a new root is created, and points to the
-// old root, and incorporates a new child, which proceeds to be
-// filled up (link) to more leaves. In all cases, the Data (chunks)
-// is stored only at the leaves, with the rest of nodes only
-// storing links to their children.
+// Package balanced provides methods to build balanced DAGs. In a
+// balanced DAG, nodes are added to a single root until the maximum
+// number of links is reached (with leaves being at depth 0). Then, a new
+// root is created to point to the old root (incorporating it as its
+// first child), which proceeds to be filled up (link) to more nodes. In
+// all cases, the data (chunk) is stored only at the leaves, with the
+// rest of (internal) nodes only storing links to their children and
+// sizes of the data stored under them.
```

Comment:
> Do we mirror the structure so the whole graph has the same depth

Great question! You *mirror* what you already had, and that term definitely needs to go in the description (yes, one nice graph can clarify all of that), the *balanced* adjective already implies it (the 1st graph is balanced and the 2nd isn't) but may not be clear enough. The problem with the old explanation:

> Then, a new root is created, and points to the old root, and incorporates a new child, which proceeds to be filled up (link) to more leaves.

Is that it is correct for the first iteration, but saying that a root's child (or any child for that matter) with be filled with leaves sound like:

```
                                  +-------------+
                                  |   Root 4    |
                                  +-------------+
                                        |       
                          +-------------+ - - - -- - -+
                          |                           |
                    +-------------+            +=============+
                    |   Root 3    |            |   Chunk 4   |
                    +-------------+            +=============+
                          |       
            +-------------+-------------+
            |                           |
     +-------------+             +=============+
     |   Root 1    |             |   Chunk 3   |
     +-------------+             +=============+
            |            
     +------+------+     
     |             |     
+=========+   +=========+
| Chunk 1 |   | Chunk 2 |
+=========+   +=========+
```
which actually is the graph you did for the new explanation that talks about nodes (generic, not internal or leaves, because that depends on the depth), I'm interested in why you reached that conclusion. But the new explanation may also be too plain, let me iterate a bit on your version and ping you back. I think the *mirror* term is the key (and the graph, no question), that's why I talk about a B-tree in the package doc, but that may not be the most appropriate data structure.

Reply:
> This graph is actually the one you associated with the new explanation that talks about nodes… I'm interested in why you reached that conclusion.

Ah, sorry if I didn’t make that clear. Here’s the sticky bit:

> Then, a new root is created to point to the old root (incorporating it as its first child), which proceeds to be filled up

So according to that sentence, the new root is the the thing that “gets filled up (link) to more nodes.” That’s technically true, but since it doesn’t say anything about creating intermediary parents/roots/internals/whatever, it sounds like we are filling the rest of *the new root* with leaf nodes.

> I think the mirror term is the key

I do think “mirror” is useful, BUT I also wrote all of this comment before getting to the bottom where you wrote “Balanced DAGs are generalistic DAGs in which all leaves are at the same distance from the root,” which totally resolved the question for me. Just moving that from the bottom to be directly below this introductory paragraph (or below a diagram which is below the intro paragraph) would also be plenty clear, I think.

Reply:
```

                                               +-------------+
                                               |   Root 6    |
                                               +-------------+
                                                     |
                          +--------------------------+----------------------------+
                          |                                                       | 
                    +-------------+                                         +-------------+
                    |   Root 3    |                                         |   Root 5    |
                    +-------------+                                         +-------------+
                          |                                                       |        
            +-------------+-------------+                           +-------------+ - - - -
            |                           |                           |              
     +-------------+             +-------------+             +-------------+       
     |   Root 1    |             |   Root 2    |             |   Root 4    |       
     +-------------+             +-------------+             +-------------+       
            |                           |                           |              
     +------+------+             +------+------+             +------+ - - - 
     |             |             |             |             |              
+=========+   +=========+   +=========+   +=========+   +=========+  
| Chunk 1 |   | Chunk 2 |   | Chunk 3 |   | Chunk 4 |   | Chunk 5 |  
+=========+   +=========+   +=========+   +=========+   +=========+  

```

I'm just realizing that most of the graphs have too many `Root` nodes, if by that we're meaning a node that was at some point a root node created by `Layout` (and has since become and internal node when the DAG grew in depth). `Root 1` is correct but `Root 3` is actually `Root 2` (if we're numbering in chronological order), the one drawn as 2 was never actually a root node, just an internal node generated by `fillNodeRec` function at the time that (what you're drawing as) `Root 3` was the root node and `Root 1` was just an internal node (`fillNodeRec` was mirroring that internal node that used to be a root). Is that clear? Maybe we could add a more descriptive legend, e.g.,

```
                    +-------------+              
                    |   Root 2    |              
                    +-------------+              
                          |                      
            +-------------+---------+            
            |                       |            
     +---------------+        +--------------+   
     |   Internal 1  |        |   Internal 3 |   
     |   (ex Root 1) |        |              |   
     +---------------+        +--------------+   
```

(BTW, have you been doing all these graphs manually? I just did a partial redrew of one of yours and became exhausted, we definitely need a tool for this.)

Reply:
> most of the graphs have too many Root nodes, if by that we're meaning a node that was at some point a root node created by Layout… the one drawn as 2 was never actually a root node

Ah, I didn’t mean anything that specific by root; only that it is the root of some sub-graph. I was trying not to *change* the name of a node between two diagrams that represent the graphs at different points in time and also trying not to introduce too much complexity. If you found that confusing, though, maybe it was a bad approach.

Maybe just “node?” (Still keep the leaves as “chunk,” though.) Not really sure.

```
     +-------------+
     |   Node 1    |
     +-------------+
            |
     +------+------+
     |             |
+=========+   +=========+   + - - - - + 
| Chunk 1 |   | Chunk 2 |   | Chunk 3 |
+=========+   +=========+   + - - - - +

                        ↯

                    +------------+
                    |   Node 2   |
                    +------------+
                          |       
            +-------------+-------------+
            |                           |
     +-------------+             +-------------+
     |    Node 1   |             |    Node 3   |
     +-------------+             +-------------+
            |                           |       
     +------+------+             +------+ - - -+
     |             |             |             |
+=========+   +=========+   +=========+   + - - - - + 
| Chunk 1 |   | Chunk 2 |   | Chunk 3 |   | Chunk 4 |
+=========+   +=========+   +=========+   + - - - - +
```

> Root 1 is correct but Root 3 is actually Root 2 (if we're numbering in chronological order)

Yeah, after reading the code, I realized this was the case. I was sort of numbering them chronologically, but it wasn’t clear from the description which was the right order (I don’t think the description actually needs to make it clear — it’s not important to a consumer of the package and would add unnecessary complex detail to a package-level explanation). It would be a good idea to renumber the chronologically anyway, though :)

> have you been doing all these graphs manually?

Yes, I have probably drawn too many weird ASCII diagrams at this point in my life :P

For diagrams like this, I’m not sure they’re so complex that finding and making a tool do what you want is really all that worthwhile (any more complex and it is, though). I will say that an editor with rectangular selections (TextMate, VS Code, etc.) is *essential,* though. You have to be able to pick up chunks of the graph and move them around, otherwise manipulating these graphs is insanely painful.

Reply:
> Ah, sorry if I didn’t make that clear. Here’s the sticky bit:
> 
> Then, a new root is created to point to the old root (incorporating it as its first child), which proceeds to be filled up
> 
> So according to that sentence, the new root is the the thing that “gets filled up (link) to more nodes.” That’s technically true, but since it doesn’t say anything about creating intermediary parents/roots/internals/whatever, it sounds like we are filling the rest of the new root with leaf nodes.
> 
> (The old explanation was clearer for me on this point because it says “incorporates a new child, which proceeds to be filled up.” It’s explicit about it being the new child that is filled.)

Yes, I see it now, thanks. I think I will also reserve the term *fill* for the parent of the leaf nodes to make that extra clear (although any node can be filled, but just for the sake of the explanation).



> Just moving that from the bottom to be directly below this introductory paragraph (or below a diagram which is below the intro paragraph) would also be plenty clear, I think.

Ok. I think the mirror term can be especially useful in the `Layout` explanation when we're talking about how DAGs are built, and leave the *balanced* term for the presentation of the full DAG as you said.

Reply:
> It would be a good idea to renumber the chronologically anyway, though :)

Yes, let's use correct numbering without explaining them (at least in the intro, maybe explain them in the `Layout` function, which might be read more by an interested developer than by the consumer of the package).

> For diagrams like this, I’m not sure they’re so complex that finding and making a tool do what you want is really all that worthwhile (any more complex and it is, though). I will say that an editor with rectangular selections (TextMate, VS Code, etc.) is essential, though. You have to be able to pick up chunks of the graph and move them around, otherwise manipulating these graphs is insanely painful.

Yes, I use the Sublime editor for that. My worry is not complexity but time consumption, as of now I haven't seen *any* diagrams like these in the code base, and if we want to incentive their incorporation we need to make it damn easy for the developer to do it, otherwise it won't be done; and we need to find a tool for that, so the developer can focus exclusively (or as much as possible) in what is being done (the layout of the DAG) instead how it will be done (the editing part, not everyone is an ascii ninja like you :)).